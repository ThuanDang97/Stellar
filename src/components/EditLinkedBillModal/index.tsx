import React, {
  memo,
  FC,
  useMemo,
  useState,
  useCallback,
  useEffect,
  FormEvent,
} from 'react'
import { useTheme } from 'styled-components'

// Props type
import { EditLinkedBillProps } from '@self-types/components/EditLinkedBillModal.props'

// Components
import Button from '@components/Button'
import Input from '@components/Input'
import Avatar from '@components/Avatar'
import Picker from '@components/Picker'
import Select from '@components/Select'
import Modal from '@components/Modal'
import { ViewStyled } from '@components/styled-components'
import AutoComplete from '@components/AutoComplete'
import {
  TitlePageStyled,
  TitlePopupStyled,
} from '@components/Title/TitleStyled'

// Constants
import {
  LIST_BILLER,
  LIST_PAYMENTS_BY_MONTH,
  MAX_LENGTH_INPUT,
  TITLE_EDIT_LINKED_BILL,
} from '@constants/index'

// Utils
import { findItemByValue, formatStringToCurrency } from '@utils/index'
import ConnectAccountForm from '@components/ConnectAccountForm'

const initBillInfo = { title: '', imgUrl: '' }

const EditLinkedBillModal: FC<EditLinkedBillProps> = ({
  editLinkedBillForm,
  appName,
  appImage,
  listBankNumber,
  status,
  isOpen,
  onClose,
  billItem,
  setBillItem,
  onSubmitBill,
  handleChangeAmount,
  handleChangeBankNumber,
  handleChangePayByDay,
  handleChangePayByMonth,
  handleSubmit,
  isConnectAccount,
  setIsConnectAccount,
  newAppImage,
  newAppName,
  errorMessageConnectAccount,
  loginConnectAccount,
  handleChangeEmail,
  handleChangePassword,
  handleSubmitConnectAccount,
}) => {
  const theme = useTheme()
  const [isSelectBill, setIsSelectBill] = useState(false)
  const [billInfo, setBillInfo] = useState(initBillInfo)

  const isDisabled = status === 'paid'

  const selectedBill = findItemByValue({
    data: LIST_BILLER as { id: string; src: string; title: string }[],
    value: billItem,
    key: 'title',
  })

  useEffect(() => {
    if (selectedBill && isSelectBill) {
      setBillInfo({ title: selectedBill.title, imgUrl: selectedBill.src })
    }
  }, [billItem, isSelectBill, selectedBill])

  const onCloseSelectBiller = useCallback(() => {
    setIsSelectBill(false)
    setBillItem('')
  }, [setBillItem])

  const handleCloseConnectAccountForm = useCallback(() => {
    setIsConnectAccount(false)
  }, [setIsConnectAccount])

  const onOpenSelectBiller = useCallback(() => setIsSelectBill(true), [])

  const onCloseModal = useCallback(() => {
    setIsSelectBill(false)
    setIsConnectAccount(false)
    setBillInfo(initBillInfo)

    if (onClose) {
      onClose()
    }
  }, [onClose, setIsConnectAccount])

  const handleSubmitAccount = useCallback(
    (e: FormEvent<HTMLFormElement>) => {
      handleSubmitConnectAccount(e)

      setIsSelectBill(false)
      setBillInfo(initBillInfo)
    },
    [handleSubmitConnectAccount],
  )

  // Render avatar and title confirm detail form
  const renderHeaderForm = useMemo(() => {
    return (
      <ViewStyled
        display="flex"
        justifyContent="center"
        flexDirection="column"
        minWidth={theme.metrics.width.xl}
      >
        <Avatar
          imageUrl={billInfo.imgUrl ? billInfo.imgUrl : appImage}
          width={80}
          height={80}
        />
        <ViewStyled display="flex">
          <TitlePopupStyled
            fontSize={theme.typography.fontSize.md}
            color={theme.colors.black}
            lineHeight={theme.typography.lineHeight.md}
            fontWeight={theme.typography.fontWeight.bold}
            display="flex"
          >
            {`${billInfo.title ? billInfo.title : appName || ''} Account`}
          </TitlePopupStyled>

          {!isSelectBill && !isDisabled && (
            <Button
              onClick={onOpenSelectBiller}
              imgUrl="/icons/pencil-square.svg"
              data-testid="iconEdit"
              variant="icon"
              size="noPadding"
              width={theme.metrics.dimensions.base}
              height={theme.metrics.dimensions.base}
              mLeft={theme.metrics.dimensions.xs}
              mTop={3}
              titleHover="Change Biller"
            />
          )}
        </ViewStyled>
      </ViewStyled>
    )
  }, [
    theme.metrics.width.xl,
    theme.metrics.dimensions.base,
    theme.metrics.dimensions.xs,
    theme.typography.fontSize.md,
    theme.typography.lineHeight.md,
    theme.typography.fontWeight.bold,
    theme.colors.black,
    billInfo.imgUrl,
    billInfo.title,
    appImage,
    appName,
    isSelectBill,
    isDisabled,
    onOpenSelectBiller,
  ])

  return (
    <Modal
      isOpen={isOpen}
      onClose={onCloseModal}
      minWidth={theme.metrics.width.xl}
    >
      {isConnectAccount ? (
        <ConnectAccountForm
          isEditModal
          email={loginConnectAccount.email}
          password={loginConnectAccount.password}
          handleChangeEmail={handleChangeEmail}
          handleChangePassword={handleChangePassword}
          handleSubmit={handleSubmitAccount}
          errorMessage={errorMessageConnectAccount}
          nameApplication={newAppName}
          imageApplication={newAppImage}
          onCloseConnectAccountForm={handleCloseConnectAccountForm}
        />
      ) : (
        <>
          <TitlePageStyled
            fontFamily="AdobeCleanExtraBold"
            color={theme.colors.black}
            fontSize={theme.typography.fontSize.xlg}
            pBottom={theme.metrics.dimensions.xs}
          >
            {TITLE_EDIT_LINKED_BILL}
          </TitlePageStyled>
          <ViewStyled
            data-testid="editLinkedBillForm"
            as="form"
            onSubmit={isSelectBill ? onSubmitBill : handleSubmit}
            noValidate
            padding={theme.metrics.dimensions.base}
            textAlign="center"
            bgColor={theme.colors.white}
            width={theme.metrics.width.full}
          >
            {renderHeaderForm}

            {isSelectBill ? (
              <>
                <AutoComplete
                  billerItem={billItem}
                  setBillerItem={setBillItem}
                />

                <ViewStyled
                  width={theme.metrics.width.full}
                  display="flex"
                  justifyContent="flex-end"
                  gap={theme.metrics.dimensions.base}
                  mTop={theme.metrics.dimensions.xxl}
                >
                  <Button
                    bgColorHover={theme.colors.ironLight}
                    onClick={onCloseSelectBiller}
                    title="Cancel"
                    color={theme.colors.shark}
                    variant="dark"
                    btnWidth="25%"
                  />
                  <Button
                    type="submit"
                    title="Continue"
                    btnWidth="28%"
                    disabled={!selectedBill}
                  />
                </ViewStyled>
              </>
            ) : (
              <>
                <Input
                  placeholder="&nbsp;Amount"
                  type="text"
                  name="amount"
                  mTop={theme.metrics.dimensions.tiny}
                  width={13}
                  mRight={-1}
                  value={formatStringToCurrency(editLinkedBillForm.amount)}
                  onChange={handleChangeAmount}
                  iconUrl={
                    isDisabled
                      ? '/icons/currency-dollar-disable'
                      : '/icons/currency-dollar'
                  }
                  readOnly={isDisabled}
                  maxLength={MAX_LENGTH_INPUT}
                />
                <Select
                  fontSize={theme.typography.fontSize.base}
                  defaultValue={editLinkedBillForm.bankName}
                  width={theme.metrics.width.full}
                  handleSelect={handleChangeBankNumber}
                  listOption={listBankNumber}
                  readOnly={isDisabled}
                  color={
                    isDisabled ? theme.colors.spanishGray : theme.colors.shark
                  }
                  borderColor={
                    isDisabled ? theme.colors.spanishGray : theme.colors.shark
                  }
                  {...{ cursor: isDisabled ? 'mouse' : 'pointer' }}
                />
                <Picker
                  onChange={handleChangePayByDay}
                  isMinDate
                  isMaxDate
                  defaultDate={editLinkedBillForm.draftDate}
                  disabled={isDisabled}
                  borderColor={
                    isDisabled ? theme.colors.spanishGray : theme.colors.cello
                  }
                  mTop={theme.metrics.dimensions.lg}
                  mBottom={theme.metrics.dimensions.sm}
                />
                <Select
                  fontSize={theme.typography.fontSize.base}
                  defaultValue={editLinkedBillForm.payPeriod}
                  width={theme.metrics.width.full}
                  handleSelect={handleChangePayByMonth}
                  listOption={LIST_PAYMENTS_BY_MONTH}
                />

                <Button
                  mTop={theme.metrics.dimensions.xxl}
                  {...{ cursor: 'pointer' }}
                  type="submit"
                  title="Confirm & Link Bill"
                  btnWidth={theme.metrics.width.full}
                  size="large"
                  disabled={!editLinkedBillForm.amount}
                />
              </>
            )}
          </ViewStyled>
        </>
      )}
    </Modal>
  )
}

export default memo(EditLinkedBillModal)
