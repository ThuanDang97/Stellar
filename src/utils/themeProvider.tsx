import { ThemeProvider } from 'styled-components'
import { render } from '@testing-library/react'
import { theme } from '@themes/index'

export const renderWithTheme = (component: React.ReactNode) => {
  return render(<ThemeProvider theme={theme}>{component}</ThemeProvider>)
}
