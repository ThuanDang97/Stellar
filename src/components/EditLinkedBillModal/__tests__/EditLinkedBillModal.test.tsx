import { cleanup } from '@testing-library/react'

// Utils
import { renderWithTheme } from '@utils/themeProvider'

// Components
import EditLinkedBillModal from '@components/EditLinkedBillModal'

// Types
import { EditLinkedBillFormProps } from '@self-types/components/EditLinkedBillModal.props'

describe('ConfirmDetailForm Component', () => {
  const editLinkedBillForm: EditLinkedBillFormProps = {
    id: '123',
    amount: '15.99',
    bankName: 'Chase',
    draftDate: 25,
    payPeriod: '3',
  }
  const props = {
    editLinkedBillForm,
    handleChangeAmount: jest.fn(),
    handleChangeBankNumber: jest.fn(),
    handleChangePayByDay: jest.fn(),
    handleChangePayByMonth: jest.fn(),
    handleSubmit: jest.fn(),
    nameApplication: 'Netflix',
    imageApplication: '/images/netflix.png',
    listBankNumber: [],
    listNumberPaymentsMonth: [],
    status: 'PAID',
    isOpen: true,
    billItem: '',
    setBillItem: jest.fn(),
    onSubmitBill: jest.fn(),
    isConnectAccount: false,
    setIsConnectAccount: jest.fn(),
    loginConnectAccount: { email: '', password: '' },
    handleSubmitConnectAccount: jest.fn(),
  }
  beforeEach(() => {
    renderWithTheme(<EditLinkedBillModal {...props} />)
  })

  afterEach(() => {
    cleanup()
  })

  it('matches snapshot', () => {
    jest.useFakeTimers()
    jest.setSystemTime(new Date('2022-08-30'))

    const { asFragment } = renderWithTheme(<EditLinkedBillModal {...props} />)

    expect(asFragment()).toMatchSnapshot()

    jest.useRealTimers()
  })
})
