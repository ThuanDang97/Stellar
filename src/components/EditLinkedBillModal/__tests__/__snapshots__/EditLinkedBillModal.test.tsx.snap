// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmDetailForm Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="ViewStyled-sc-rqe5sa-0 ModalStyled__BackDropStyled-sc-d5vki3-1 fTsvHV cnMCGd"
    data-testid="backdrop"
  >
    <div
      class="ViewStyled-sc-rqe5sa-0 ViewStyled__BoxShadowStyled-sc-rqe5sa-2 ModalStyled-sc-d5vki3-0 eUUuqy hTNOqC hmXhmG"
    >
      <div
        class="ModalStyled__ImageWrapper-sc-d5vki3-2 eSAkbC"
        data-testid="close-icon"
      >
        <span
          style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
        >
          <img
            alt="close-icon"
            data-nimg="fill"
            decoding="async"
            src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
            style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
          />
          <noscript />
        </span>
      </div>
      <span
        class="TextStyled-sc-9xia41-0 TitleStyled-sc-t58pjn-0 TitleStyled__TitlePageStyled-sc-t58pjn-1 dRSlxp WstJK ilgUUT"
        color="#000"
        font-family="AdobeCleanExtraBold"
        font-size="36"
      >
        Edit Linked Bill
      </span>
      <form
        class="ViewStyled-sc-rqe5sa-0 giGcYc"
        data-testid="editLinkedBillForm"
        novalidate=""
        width="100%"
      >
        <div
          class="ViewStyled-sc-rqe5sa-0 fldtEt"
          display="flex"
        >
          <div
            class="ViewStyled-sc-rqe5sa-0 gjRQva"
            cursor="pointer"
            height="80"
            width="80"
          >
            <span
              class="AvatarStyled-sc-1ppokaq-1 bOiqfJ"
              data-testid="avatarDefault"
            />
          </div>
          <div
            class="ViewStyled-sc-rqe5sa-0 knVWgS"
            display="flex"
          >
            <span
              class="TextStyled-sc-9xia41-0 TitleStyled-sc-t58pjn-0 TitleStyled__TitlePopupStyled-sc-t58pjn-2 ejoqyt kyLpEc iyHIHp"
              color="#000"
              display="flex"
              font-size="20"
              font-weight="700"
            >
               Account
            </span>
            <button
              class="ButtonStyled-sc-1ys9v52-0 jLTkdk"
              data-testid="iconEdit"
              title="Change Biller"
            >
              <span
                class="ButtonStyled__ButtonIconStyled-sc-1ys9v52-1 DjHyU"
              >
                <div
                  class="ViewStyled-sc-rqe5sa-0 efewJe"
                  display="inline-block"
                  height="15"
                  width="15"
                >
                  <span
                    style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
                  >
                    <img
                      data-nimg="fill"
                      decoding="async"
                      src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                      style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
                    />
                    <noscript />
                  </span>
                </div>
              </span>
            </button>
          </div>
        </div>
        <div
          class="ViewStyled-sc-rqe5sa-0 InputStyled__WrapperInputStyled-sc-9gj86m-1 gFYjgc brcjID"
          color="#203F59"
          data-testid="inputIcon"
          display="block"
        >
          <div
            class="ViewStyled-sc-rqe5sa-0 InputStyled__WrapperIconStyled-sc-9gj86m-2 bmWIbW kQnIvY"
            width="13"
          >
            <span
              style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
            >
              <img
                alt="icon"
                data-nimg="fill"
                decoding="async"
                src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
              />
              <noscript />
            </span>
          </div>
          <input
            class="InputStyled-sc-1k09cqb-0 InputStyled__InputItemStyled-sc-9gj86m-0 hnxmhM NDCRX"
            data-testid="iconInput"
            maxlength="12"
            name="amount"
            placeholder="Â Amount"
            type="text"
            value="15.99"
          />
        </div>
        <div
          class="ViewStyled-sc-rqe5sa-0 fTsvHV"
        >
          <div
            class="ViewStyled-sc-rqe5sa-0 ViewStyled__BorderWrapperStyled-sc-rqe5sa-4 eJgyhf iTrmsK"
            data-testid="select"
            pointer-events="auto"
            width="100%"
          >
            <div
              class="ViewStyled-sc-rqe5sa-0 SelectStyled__WrapperSelectStyled-sc-4qihay-1 fTsvHV kUJXAU"
            >
              <div
                class="ViewStyled__DividerStyled-sc-rqe5sa-5 cmkogR"
                width="100%"
              >
                <span
                  class="TextStyled-sc-9xia41-0 dItzST"
                  color="#242A30"
                  data-testid="select-value"
                  font-size="14"
                />
                <span
                  class="SelectStyled__SelectIconStyled-sc-4qihay-0 gLdeRH"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="ViewStyled-sc-rqe5sa-0 ViewStyled__BorderWrapperStyled-sc-rqe5sa-4 dyvxXI eWayqq"
          data-testid="select"
          width="100%"
        >
          <div
            class="ViewStyled__DividerStyled-sc-rqe5sa-5 eRzaft"
          >
            <div
              class="ViewStyled-sc-rqe5sa-0 InputStyled__WrapperInputStyled-sc-9gj86m-1 gtErIS llBddb"
              display="flex"
            >
              <div
                class="react-datepicker-wrapper"
              >
                <div
                  class="react-datepicker__input-container"
                >
                  <input
                    class="PickerStyled__DateStyled-sc-1ns5yoz-0 golTem"
                    placeholder="MM/DD/YYYY"
                    type="text"
                    value="Pay by the 25th"
                  />
                </div>
              </div>
              <span
                class="SelectStyled__SelectIconStyled-sc-4qihay-0 kogIxx"
              />
            </div>
          </div>
        </div>
        <div
          class="ViewStyled-sc-rqe5sa-0 fTsvHV"
        >
          <div
            class="ViewStyled-sc-rqe5sa-0 ViewStyled__BorderWrapperStyled-sc-rqe5sa-4 eJgyhf iTrmsK"
            data-testid="select"
            pointer-events="auto"
            width="100%"
          >
            <div
              class="ViewStyled-sc-rqe5sa-0 SelectStyled__WrapperSelectStyled-sc-4qihay-1 fTsvHV kUJXAU"
            >
              <div
                class="ViewStyled__DividerStyled-sc-rqe5sa-5 cmkogR"
                width="100%"
              >
                <span
                  class="TextStyled-sc-9xia41-0 dItzST"
                  data-testid="select-value"
                  font-size="14"
                >
                  Pay 3 months
                </span>
                <span
                  class="SelectStyled__SelectIconStyled-sc-4qihay-0 loNHhu"
                />
              </div>
            </div>
          </div>
        </div>
        <button
          class="ButtonStyled-sc-1ys9v52-0 hFEKXM"
          cursor="pointer"
          data-testid="button"
          type="submit"
        >
          Confirm & Link Bill
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
